### DISCUSSION TITLE
When using `useRouter`, is it normal for the `router.query` object to be empty on 1st render on the client?

### DISCUSSION THREAD

**Discussion by cbdeveloper** [vercel/next.js#24659] [Help] [answered]
I have a page component that does the following:

```
/pages
  /search
    index.tsx
```

My page component gets its data on the client. So I'm not using any data methods on my `search/index.tsx` file.

The search url will use query string. So it should be something like:

`/search?q=search+slug` for example.

This is my `index.tsx` file:

```
import React from "react";
import SearchContainer from "@app/pages/search/SearchContainer";

interface SearchRoute {}

const SearchRoute: React.FC<SearchRoute> = (props) => {

  return(
    <SearchContainer/>
  );
};

export default React.memo(SearchRoute);
```

And this is my `SearchContainer`. I'm using redux to manage the state for that page.

```
const SearchContainer: React.FC<SearchContainer> = (props) => {

  const { status } = usePageSearchState();
  const dispatch = useDispatch();
  const router = useRouter();
  const query = router.query as { q: string };
  const queryString = query.q;

  console.log("SEARCH CONTAINER");
  console.log(`queryString: ${queryString}`);

  useEffect(() => {
    dispatch(THUNK.LOAD(queryString));
    return () => { dispatch(ACTION.RESET_STATE()); };
  },[dispatch,queryString]);

  return(
    <AsyncContainer status={status}>
      <SearchPage/>
    </AsyncContainer>
  );
};
```

What is going on:

- I need to read `router.query.q` to get the `search slug` and dispatch the `THUNK.LOAD(queryString)`.
- In theory, I should fire the `THUNK.LOAD` only once per `search slug` value.
- But I noticed that the `THUNK.LOAD` was being fired twice.
- That's why I've added those `console.log` to see what was going on with the `router.query.q` value.

And for my surprise, this is getting logged:

![image](https://user-images.githubusercontent.com/43407798/116712718-0e1aec00-a9cc-11eb-8094-924a469e56ed.png)

And then:

![image](https://user-images.githubusercontent.com/43407798/116712741-13783680-a9cc-11eb-977c-0c4b180ca350.png)

It means that on the 1st run `router.query` is probably empty `{}`, given the fact that `router.query.q` is coming back as undefined.

This only happens when I fully refresh the app in that URL: `/search?q=search+slug` (like hitting F5 on that page). If the app has already been loaded in a different page and I do a search that leads me do that page, then the code behaves normally and I don't get the `undefined` read. I get the `router.query` as `{ q: "search slug" }` right away.

Is this normal?

Is there any reason for the `router.query` to come back as `{}` on the first render **on the client** ?

I mean, the [documentation](https://nextjs.org/docs/api-reference/next/router#router-object) mentions that it should be empty during prerendering if the page does not have data fetching methods. But does not mention anything about its behavior on the client.

![image](https://user-images.githubusercontent.com/43407798/116713619-dfe9dc00-a9cc-11eb-9398-519fa8d425b9.png)

I wasn't expecting that behavior at all. Is this normal?

I'm also asking this on Stack Overflow [link](https://stackoverflow.com/questions/67332108/next-js-router-is-returning-query-parameters-as-undefined-on-first-render).

**Comment by techrome** [comment]
I'm not sure but maybe you should remove `React.memo` from this page and try if it works without it? `export default React.memo(SearchRoute);`

Also what version of `next` are you using? There are sometimes minor bugs that get fixed in newer versions, so if you can update to `10.2.0` version (latest) maybe it will help?

**Reply by cbdeveloper** [reply]
I'm using `10.1.3`.

I was exploring the source code from `useRouter()`.

It initially gets the `router.query` value from `window.__NEXT_DATA__` and that value is actually empty `{}` on `__NEXT_DATA__`, which is expected, since I ain't got no data fetching methods for that page.

What I wasn't able to find out is why would it wait for a second pass to populate the `router.query` instead of getting it from the `window.location` object right away.

Do you think it's a version issue?

**Reply by cbdeveloper** [reply]
The fact that It's `undefined` on the first run is problematic, because I can't be sure if I the user hasn't entered a valid query, or if it's `undefined` just because I'm on the first run. For example:

User hits `/search?q=XXX`

```
const router = useRouter();
const { q  }= router.query;

useEffect(() => {
  if (q)
    doSearch(q);
},[q]);
```

The code above should work fine, because `q` will be `undefined` first, and then it will become `XXX` and my search will run.

But what if the user hits: `/search?q=`  or maybe just `/search` without any query?

Then that code will not work, because the final value for `q` will actually be `""` or `undefined` and my code won't be aware of that. As a workaround I would have to keep track if the code is on the 1st run or not, which is not ideal at all.

Does this make sense to you guys?

I mean, I can always make that page as `SSR` by adding a `getServeSideProps` to it, but then I'll lose cache and also I won't be able to show a spinner, while my user is waiting for the search to complete.

**Comment by cbdeveloper** [comment, marked as answer]
Ok. Just found out what is the solution to my problem:

From: https://nextjs.org/docs/api-reference/next/router#router-object

![image](https://user-images.githubusercontent.com/43407798/116863925-ecee1180-abfe-11eb-8814-b927e5649936.png)

This is what happens to the `router.query` on client when you hit `/search?q=XXX`

1st render

![image](https://user-images.githubusercontent.com/43407798/116864092-363e6100-abff-11eb-83c5-8f617130a7b2.png)

Subsequent renders

![image](https://user-images.githubusercontent.com/43407798/116864111-3cccd880-abff-11eb-904b-50c8fbd6a722.png)

**Conclusion**

The fact that `router.query` is not populated on the client (for SSG pages) on the first run is a design implementation detail. And you can monitor whether it's has been populated or not by checking the `router.isReady` boolean property.
